### PR TITLE
Add an annotation for @Config elements which will automatically create a slider control

### PR DESCRIPTION
**Pull request by AlexiyOrlov**
My first PR; adds an annotation which, if placed on numeric fields, will cause a creation of slider control in the associated config GUI. Marking other fields will cause nothing.
Makes use of `GuiConfigEntries.NumberSliderEntry`.

### CODE REVIEW

File: src/main/java/net/minecraftforge/fml/client/config/IConfigElement.java [hunk 1]
```diff
@@ -186,4 +186,9 @@
      * [Property] Gets a Pattern object used in String property input validation.
      */
     Pattern getValidationPattern();
+
+    /**
+     * @return true if this element is going to have a slider attached
+     */
+    boolean hasSlidingControl();
```
Comment (Alpvax) [inline src/main/java/net/minecraftforge/fml/client/config/IConfigElement.java]: Can we make this default to false to preserve backwards compatibility?

File: src/main/java/net/minecraftforge/common/config/ConfigManager.java [hunk 2]
```diff
@@ -242,6 +242,8 @@ private static void sync(Configuration cfg, Class<?> cls, String modid, String c
             boolean requiresMcRestart = f.isAnnotationPresent(Config.RequiresMcRestart.class);
             boolean requiresWorldRestart = f.isAnnotationPresent(Config.RequiresWorldRestart.class);
 
+            boolean hasSlidingControl=f.isAnnotationPresent(SlidingOption.class);
```
Comment (iTitus) [inline src/main/java/net/minecraftforge/common/config/ConfigManager.java]: Whitespaces missing around the equals sign

File: src/main/java/net/minecraftforge/common/config/SlidingOption.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package net.minecraftforge.common.config;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Created on 7/9/18 by alexiy.
```
Comment (iTitus) [inline src/main/java/net/minecraftforge/common/config/SlidingOption.java]: Can you do the right header here

File: src/test/java/net/minecraftforge/test/SliderConfigElementTest.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        @SlidingOption
+        public static String string="I don't slide";
+
+
+    }
+
+    @SubscribeEvent
+    public static void optionsChanged(ConfigChangedEvent.OnConfigChangedEvent configChangedEvent)
+    {
+        if(configChangedEvent.getModID().equals(ID))
```
Comment (iTitus) [inline src/test/java/net/minecraftforge/test/SliderConfigElementTest.java:64]: Missing space after if

File: src/test/java/net/minecraftforge/test/SliderConfigElementTest.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+import net.minecraftforge.common.config.Config;
+import net.minecraftforge.common.config.ConfigManager;
+import net.minecraftforge.common.config.SlidingOption;
+import net.minecraftforge.fml.client.event.ConfigChangedEvent;
+import net.minecraftforge.fml.common.Mod;
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
+
+/**
+ * Created on 7/9/18 by alexiy.
```
Comment (iTitus) [inline src/test/java/net/minecraftforge/test/SliderConfigElementTest.java]: Header...

**Comment by kierajreed** [comment]
The license in [this file](https://github.com/MinecraftForge/MinecraftForge/blob/1.12.x/src/main/java/net/minecraftforge/common/chunkio/ChunkIOThreadPoolExecutor.java) is incorrect. I think that if you rebase and fix it the Travis build will work fine.

**Comment by AlexiyOrlov** [comment]
The travis build succeeded?

**Comment by kierajreed** [comment]
mezz fixed the license and reran the build.

File: src/main/java/net/minecraftforge/common/config/SlidingOption.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * A field marked with this annotation will have a slider control attached in the config UI
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.FIELD)
+public @interface SlidingOption {
```
Comment (thilokru) [inline src/main/java/net/minecraftforge/common/config/SlidingOption.java]: Why isn't this part of the @interface Config class?

**Comment by MajorTuvok** [comment]
Go away stale - this is a nice feature...

**Comment by AlexiyOrlov** [comment]
I hope this will be merged into 1.12.2 branch? Really waiting for this feature to use myself...

File: src/test/resources/mcmod.info [hunk 7]
```diff
@@ -1,16 +1,8 @@
 [
 {
-  "modid": "config_test",
```
Comment (tterrag1098) [inline src/test/resources/mcmod.info]: This breaks the other test, you must define both mods here.

**Review by tterrag1098 — CHANGES_REQUESTED** [review]
Otherwise looks good, just fix this minor issue

File: src/test/resources/mcmod.info (hunk 7 above)
Comment (AlexiyOrlov) [reply src/test/resources/mcmod.info]: I didn't realize that "mcmod"s are Json files, too, hah. I learned it now, thanks.

**Review by tterrag1098 — APPROVED** [review]

File: src/main/java/net/minecraftforge/common/config/ConfigManager.java [hunk 8]
```diff
@@ -261,6 +263,8 @@ private static void sync(Configuration cfg, Class<?> cls, String modid, String c
                         {
                             Property property = property(cfg, wrapper.getCategory(), suffix, propType, adapt.isArrayAdapter());
 
+                            property.setHasSlidingControl(hasSlidingControl);
```
Comment (thilokru) [inline src/main/java/net/minecraftforge/common/config/ConfigManager.java]: Setting up field properties like requiresMcRestart, or here, hasSlidingControl, is handled via IFieldWrapper.setupConfiguration(). This generalizes the setup.
Please consider moving it to that method.

**Review by thilokru — CHANGES_REQUESTED** [review]
For setting up boolean values similar to your new one, there already exists a mechanism you should modify.

File: src/main/java/net/minecraftforge/common/config/IFieldWrapper.java [hunk 9]
```diff
@@ -68,7 +68,7 @@
      * @param reqMCRestart True, if a change in this property requires a restart of Minecraft
      * @param reqWorldRestart True, if the world needs to be reloaded after changes to this property
      */
-    void setupConfiguration(Configuration cfg, String desc, String langKey, boolean reqMCRestart, boolean reqWorldRestart);
+    void setupConfiguration(Configuration cfg, String desc, String langKey, boolean reqMCRestart, boolean reqWorldRestart, boolean hasSlidingControl);
```
Comment (tterrag1098) [inline src/main/java/net/minecraftforge/common/config/IFieldWrapper.java]: Let's not break binary compat here, even if it is unlikely to affect anyone.
```suggestion
    default void setupConfiguration(Configuration cfg, String desc, String langKey, boolean reqMCRestart, boolean reqWorldRestart) {
        return setupConfiguration(cfg, desc, langKey, reqMCRestart, reqWorldRestart, false);
    }
    
    void setupConfiguration(Configuration cfg, String desc, String langKey, boolean reqMCRestart, boolean reqWorldRestart, boolean hasSlidingControl);
```

**Review by tterrag1098 — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/common/config/IFieldWrapper.java [hunk 10]
```diff
@@ -68,6 +68,10 @@
      * @param reqMCRestart True, if a change in this property requires a restart of Minecraft
      * @param reqWorldRestart True, if the world needs to be reloaded after changes to this property
      */
+    default void setupConfiguration(Configuration cfg, String desc, String langKey, boolean reqMCRestart, boolean reqWorldRestart) {
+        return setupConfiguration(cfg, desc, langKey, reqMCRestart, reqWorldRestart, false);
+    }
+
```
Comment (tterrag1098) [inline src/main/java/net/minecraftforge/common/config/IFieldWrapper.java:84]: The javadocs should be on this one with an @see on the other. Or just on both.

Comment (thilokru) [reply src/main/java/net/minecraftforge/common/config/IFieldWrapper.java:84]: Additionaly, mark it for removal in the next large update. That is our chance to unify these two methods.

File: src/main/java/net/minecraftforge/common/config/IFieldWrapper.java [hunk 11]
```diff
@@ -61,17 +61,21 @@
      */
     boolean handlesKey(String key);
 
-    /**
+    /**TODO remove both methods in future
```
Comment (thilokru) [inline src/main/java/net/minecraftforge/common/config/IFieldWrapper.java]: Why should we remove both methods? We need one to do the setup. We currently have two to maintain binary compatibility. In the future, we need probably only one, but zero is one to few.

Comment (tterrag1098) [reply src/main/java/net/minecraftforge/common/config/IFieldWrapper.java]: This whole system is dead in 1.13 with the advent of the new TOML-based configuration system. So any "remove in future" comments are pointless in their entirety, this code doesn't even exist in the 1.13 codebase.

Comment (AlexiyOrlov) [reply src/main/java/net/minecraftforge/common/config/IFieldWrapper.java]: alright, I don't know what to do here anymore

Comment (thilokru) [reply src/main/java/net/minecraftforge/common/config/IFieldWrapper.java]: @tterrag1098 I designed this system to able to handle different backing configuration files without changing the interface to the mods. Therefore this infrastructure should be able to remain and maintain backwards comaptibility.

**Comment by AlexiyOrlov** [comment]
Maybe someone else will have better luck with this...

**Comment by tterrag1098** [comment]
There's no reason to close this.

**Comment by AlexiyOrlov** [comment]
Ok, what should I do here? I think requested changes have been made. 4c5ca2048c808c83d6799f10d3a54f32dace706c a72aefc

**Comment by tterrag1098** [comment]
The real javadoc (not the `@see`) should go on the most specific method, i.e. the one you added. Put the `@see` on the other one and remove the TODO entirely.

File: src/main/java/net/minecraftforge/common/config/IFieldWrapper.java [hunk 12]
```diff
@@ -61,20 +61,20 @@
...
     }
 
     /**
-     * @see IFieldWrapper#setupConfiguration(Configuration, String, String, boolean, boolean)
      *
+     * @param cfg The configuration object holding the properties
+     * @param desc The properties description
+     * @param langKey The languageKey of the property, used in GUI
+     * @param reqMCRestart True, if a change in this property requires a restart of Minecraft
+     * @param reqWorldRestart True, if the world needs to be reloaded after changes to this property
```
Comment (tterrag1098) [inline src/main/java/net/minecraftforge/common/config/IFieldWrapper.java:76]: Could you also document the new parameter?

**Review by thilokru — APPROVED** [review]

File: src/main/java/net/minecraftforge/common/config/IFieldWrapper.java (hunk 12 above)
Comment (AlexiyOrlov) [reply src/main/java/net/minecraftforge/common/config/IFieldWrapper.java:76]: Done

**Review by tterrag1098 — APPROVED** [review]

**Review by mezz — APPROVED** [review]

**Review by mezz — DISMISSED** [review]
The new `setupConfiguration` method needs to be `default` in case any mod is implementing `IFieldWrapper` on its own.
I recommend having it call the old method by default, to support that case. 

The old `setupConfiguration` method should not have a `default` implementation and it should be marked `@Deprecated` so we can remove it later. In the Forge classes implementing the interface, you can implement it so that it just call the new method like you are doing now in the interface.

Otherwise looks good and I can merge it once those are fixed.

File: src/main/java/net/minecraftforge/common/config/FieldWrapper.java [hunk 13]
```diff
@@ -174,6 +174,11 @@ public boolean handlesKey(String key)
             return key.startsWith(this.baseName);
         }
 
+        @Override
```
Comment (tterrag1098) [inline src/main/java/net/minecraftforge/common/config/FieldWrapper.java:177]: Style: brace should be on new line

File: src/main/java/net/minecraftforge/common/config/FieldWrapper.java [hunk 14]
```diff
@@ -174,6 +174,11 @@ public boolean handlesKey(String key)
...
+        public void setupConfiguration(Configuration cfg, String desc, String langKey, boolean reqMCRestart, boolean reqWorldRestart) {
+            this.setupConfiguration(cfg, desc, langKey, reqMCRestart, reqWorldRestart,false);
```
Comment (tterrag1098) [inline src/main/java/net/minecraftforge/common/config/FieldWrapper.java]: Style: space before `false`

**Review by tterrag1098 — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/common/config/FieldWrapper.java (hunk 14 above)
Comment (AlexiyOrlov) [reply src/main/java/net/minecraftforge/common/config/FieldWrapper.java]: Anything else?

**Comment by AlexiyOrlov** [comment]
I guess everyone's busy with 1.13

**Review by tterrag1098 — APPROVED** [review]

**Comment by tterrag1098** [comment]
Holidays happen. This seems ready to go now.

**Pull request merged**